src/ROOTNTupleReader.cc

```diff
@@ -0,0 +1,212 @@
+  auto valueView = m_readers[name][0]->GetView<std::vector<std::vector<T>>>(root_utils::getGPValue<T>());
+
+  auto keys = keyView(entNum);
+  auto values = valueView(entNum);
+
+  for (size_t i = 0; i < keys.size(); ++i) {
+    params.getMap<T>()[keys[i]] = values[i];
+  }
+}
+
+GenericParameters ROOTNTupleReader::readEventMetaData(const std::string& name, unsigned entNum) {
+  GenericParameters params;
+
+  readParams<int>(name, entNum, params);
+  readParams<float>(name, entNum, params);
+  readParams<double>(name, entNum, params);
+  readParams<std::string>(name, entNum, params);
+
+  return params;
+}
+
+bool ROOTNTupleReader::initCategory(const std::string& category) {
+  if (std::find(m_availableCategories.begin(), m_availableCategories.end(), category) == m_availableCategories.end()) {
+    return false;
+  }
+  // Assume that the metadata is the same in all files
+  auto filename = m_filenames[0];
+
+  auto id = m_metadata_readers[filename]->GetView<std::vector<int>>(root_utils::idTableName(category));
+  m_collectionInfo[category].id = id(0);
+
+  auto collectionName = m_metadata_readers[filename]->GetView<std::vector<std::string>>(root_utils::collectionName(category));
+  m_collectionInfo[category].name = collectionName(0);
+
+  auto collectionType = m_metadata_readers[filename]->GetView<std::vector<std::string>>(root_utils::collInfoName(category));
+  m_collectionInfo[category].type = collectionType(0);
+   
+  auto subsetCollection = m_metadata_readers[filename]->GetView<std::vector<short>>(root_utils::subsetCollection(category));
+  m_collectionInfo[category].isSubsetCollection = subsetCollection(0);
+
+  return true;
+}
+
+void ROOTNTupleReader::openFile(const std::string& filename) {
+  openFiles({filename});
+}
+
+void ROOTNTupleReader::openFiles(const std::vector<std::string>& filenames) {
+
+  m_filenames.insert(m_filenames.end(), filenames.begin(), filenames.end());
+  for (auto& filename : filenames) {
+    if (m_metadata_readers.find(filename) == m_metadata_readers.end()) {
+      m_metadata_readers[filename] = ROOT::Experimental::RNTupleReader::Open(root_utils::metaTreeName, filename);
+    }
+  }
+
+  m_metadata = ROOT::Experimental::RNTupleReader::Open(root_utils::metaTreeName, filenames[0]);
+
+  auto versionView = m_metadata->GetView<std::vector<uint16_t>>(root_utils::versionBranchName);
+  auto version = versionView(0);
+
+  m_fileVersion = podio::version::Version{version[0], version[1], version[2]};
+
+  auto edmView = m_metadata->GetView<std::vector<std::tuple<std::string, std::string>>>(root_utils::edmDefBranchName);
+  auto edm = edmView(0);
+
+  auto availableCategoriesField = m_metadata->GetView<std::vector<std::string>>(root_utils::availableCategories);
+  m_availableCategories = availableCategoriesField(0);
+
+}
+
+unsigned ROOTNTupleReader::getEntries(const std::string& name) {
+  if (m_readers.find(name) == m_readers.end()) {
```

Comment:
Can we cache this after the first calculation for a category? Otherwise constructs like the following could become costly
```cpp
for (auto i = 0u; i < reader.getEntries("events"); ++i) {
  // ... do something
}
```
Especially also because this is not even marked `const` so the compiler will most likely not even be able to hoist this to the outside of the loop.

Reply:
Done